plugins/continuous-toolbox/src/RecyclableBlockFlyoutInflater.ts

```diff
@@ -0,0 +1,193 @@
+  }
+
+  /**
+   * Creates a new block from the given block definition.
+   *
+   * @param blockDefinition The definition to create a block from.
+   * @param workspace The workspace to create the block on.
+   * @returns The newly created block.
+   */
+  override createBlock(
+    blockDefinition: Blockly.utils.toolbox.BlockInfo,
+    workspace: Blockly.WorkspaceSvg,
+  ): Blockly.BlockSvg {
+    const blockType = this.getTypeFromDefinition(blockDefinition);
+    return (
+      this.getRecycledBlock(blockType) ??
+      super.createBlock(blockDefinition, workspace)
+    );
+  }
+
+  /**
+   * Returns the type of a block from an XML or JSON block definition.
+   *
+   * @param blockDefinition The block definition to parse.
+   * @returns The block type.
+   */
+  private getTypeFromDefinition(
+    blockDefinition: Blockly.utils.toolbox.BlockInfo,
+  ): string {
+    let type: string | null | undefined;
+    if (blockDefinition['blockxml']) {
+      const xml =
+        typeof blockDefinition['blockxml'] === 'string'
+          ? Blockly.utils.xml.textToDom(blockDefinition['blockxml'])
+          : (blockDefinition['blockxml'] as Element);
+      type = xml.getAttribute('type');
+    } else {
+      type = blockDefinition['type'];
+    }
+
+    if (!type) {
+      throw new Error(
+        `Block type is not specified in block definition: ${JSON.stringify(
+          blockDefinition,
+        )}`,
+      );
+    }
+    return type;
+  }
+
+  /**
+   * Puts a previously created block into the recycle bin and moves it to the
+   * top of the workspace. Used during large workspace swaps to limit the number
+   * of new DOM elements we need to create.
+   *
+   * @param block The block to recycle.
+   */
+  private recycleBlock(block: Blockly.BlockSvg) {
+    const xy = block.getRelativeToSurfaceXY();
+    block.moveBy(-xy.x, -xy.y);
+    this.recycledBlocks.set(block.type, block);
+  }
+
+  /**
+   * Returns a block from the cache of recycled blocks with the given type, or
+   * undefined if one cannot be found.
+   *
+   * @param blockType The type of the block to try to recycle.
+   * @returns The recycled block, or undefined if one could not be recycled.
+   */
+  private getRecycledBlock(blockType: string): Blockly.BlockSvg | undefined {
+    const block = this.recycledBlocks.get(blockType);
+    this.recycledBlocks.delete(blockType);
+    return block;
+  }
+
+  /**
+   * Returns whether the given block can be recycled or not.
+   *
+   * @param block The block to check for recyclability.
+   * @returns True if the block can be recycled. False otherwise.
+   */
+  protected blockIsRecyclable(block: Blockly.Block): boolean {
+    if (!this.recyclingEnabled) {
+      return false;
+    }
+
+    if (this.recycleEligibilityChecker) {
+      return this.recycleEligibilityChecker(block);
+    }
+
+    // If the block needs to parse mutations, never recycle.
+    if (block.mutationToDom && block.domToMutation) {
+      return false;
+    }
```

Comment:
Should `saveExtraState` and `loadExtraState` not also be checked, for block definitions that don't support XML serialisation?